File: app/src/main/java/com/github/dedis/student20_pop/model/network/query/method/Unsubscribe.java
package com.github.dedis.student20_pop.model.network.query.method;

import com.github.dedis.student20_pop.model.network.query.Method;
import com.github.dedis.student20_pop.model.network.query.Query;

/**
 * Unsubscribe from a channel
 */
public final class Unsubscribe extends Query {

    public Unsubscribe(String channel, int id) {
        super(channel, id);
    }

    @Override
    public String getMethod() {
        return Method.UNSUBSCRIBE.getMethod();
    }
}

File: be1-go/internal/handler/query/unsubscribe.go
package query

import (
	"encoding/json"
	"popstellar/internal/errors"
	"popstellar/internal/handler/messagedata/root"
	"popstellar/internal/message/query/method"
	"popstellar/internal/network/socket"
	"popstellar/internal/singleton/state"
)

func handleUnsubscribe(socket socket.Socket, msg []byte) (*int, error) {
	var unsubscribe method.Unsubscribe
	err := json.Unmarshal(msg, &unsubscribe)
	if err != nil {
		return nil, errors.NewJsonUnmarshalError(err.Error())
	}

	if root.Root == unsubscribe.Params.Channel {
		return &unsubscribe.ID, errors.NewAccessDeniedError("cannot Unsubscribe from root channel")
	}

	err = state.Unsubscribe(socket, unsubscribe.Params.Channel)
	if err != nil {
		return &unsubscribe.ID, err
	}

	socket.SendResult(unsubscribe.ID, nil, nil)

	return &unsubscribe.ID, nil
}
